catkin_tools/jobs/catkin.py

```diff
@@ -0,0 +1,668 @@
+                os.symlink(source_file, dest_file)
+
+    # Load the old list of symlinked files for this package
+    if os.path.exists(devel_manifest_path):
+        with open(devel_manifest_path, 'r') as devel_manifest:
+            manifest_reader = csv.reader(devel_manifest, delimiter=' ', quotechar='"')
+            # Skip the package source directory
+            devel_manifest.readline()
+            # Read the previously-generated products
+            for source_file, dest_file in manifest_reader:
+                # print('Checking (%s, %s)' % (source_file, dest_file))
+                if (source_file, dest_file) not in products:
+                    # Clean the file or decrement the collision count
+                    logger.out('Cleaning (%s, %s)' % (source_file, dest_file))
+                    files_to_clean.append(dest_file)
+
+    # Remove all listed symlinks and empty directories which have been removed
+    # after this build, and update the collision file
+    clean_linked_files(logger, event_queue, devel_space_abs, files_that_collide, files_to_clean)
+
+    # Save the list of symlinked files
+    with open(devel_manifest_path, 'w') as devel_manifest:
+        # Write the path to the package source directory
+        devel_manifest.write('%s\n' % package_source_abs)
+        # Write all the products
+        manifest_writer = csv.writer(devel_manifest, delimiter=' ', quotechar='"')
+        for source_file, dest_file in products:
+            manifest_writer.writerow([source_file, dest_file])
+
+    return 0
+
+
+# job factories
+
+def create_catkin_build_job(context, package, package_path, dependencies, force_cmake, pre_clean):
+    """Job class for building catkin packages"""
+
+    # Package source space path
+    pkg_dir = os.path.join(context.source_space_abs, package_path)
+
+    # Package build space path
+    build_space = get_package_build_space_path(context.build_space_abs, package.name)
+
+    # Package devel space path
+    if context.isolate_devel:
+        devel_space = os.path.join(context.devel_space_abs, package.name)
+    elif context.link_devel:
+        devel_space = get_linked_devel_package_path(context.devel_space_abs, package.name)
+    else:
+        devel_space = context.devel_space_abs
+
+    # Package install space path
+    if context.isolate_install:
+        install_space = os.path.join(context.install_space_abs, package.name)
+    else:
+        install_space = context.install_space_abs
+
+    # Create job stages
+    stages = []
+
+    # Create package build space
+    stages.append(FunctionStage(
+        'mkdir',
+        makedirs,
+        path=build_space))
+
+    # Create an environment file
+    env_file_path = get_env_file_path(package, context)
+    stages.append(FunctionStage(
+        'envgen',
+        create_env_file,
+        package=package,
+        context=context,
+        env_file_path=env_file_path))
+
+    # Only use it for building if the develspace is isolated
```

Comment:
@jbohren Can you explain the motivation here? This breaks building from scratch for me, e.g.
```
# Clean workspace (just src folder)
. /opt/ros/indigo/setup.bash
catkin build
```

This fails for packages including message headers built in other packages, since `devel/include` will not be included in the header search path. Indeed, how should a package know about the shared devel space if not through the env file?

After the first build fails, I can source the then-available `devel/setup.bash` and a second build will succeed.

If I use `env_prefix = [env_file_path]` unconditionally here, the `devel/include` is added to `CPATH` and everything works on the first build.